### PR TITLE
fix: tabbed code block regex

### PR DESCRIPTION
**Pull request by kellyjosephprice**
[Demo][demo] | 🎫 [Ticket][ticket]
:---:|:---:

### 🧰  Changes

The previous regex had problems matching on poorly formatted markdown.
This one attempts to use a more precise regex to match adjacent code
blocks. It does fail the linter's unsafe-regex check, but I belive that
it's safe from 'catastrophic' backtracking. IIUC the negative lookahead
should mean there's only one way to match the nested matchers.

[demo]: #demo-app-link-to-come
[ticket]: https://app.asana.com/0/1198968777158470/1198971573743122/f

### CODE REVIEW

**Comment by rafegoldberg** [comment]
Amazing. Will check this out and review in a bit.

**Comment by kellyjosephprice** [comment]
Hmm, I guess I should have expected that would fail 10.x

File: __tests__/flavored-parsers.test.js [hunk 1]
```diff
@@ -99,4 +99,34 @@ describe('Parse RDMD Syntax', () => {
     const ast = process(mdx);
     expect(ast.children).toHaveLength(3);
   });
+
+  it('More foolish disregard for newlines.', () => {
+    const mdx = '```\nfoo\n```\nOops\n```\nbar\n```';
+    const ast = process(mdx);
+
+    expect(ast.children.map(c => c.type)).toStrictEqual(['code', 'paragraph', 'code']);
```
Comment (rafegoldberg) [inline __tests__/flavored-parsers.test.js:107]: Whoah, I hadn't thought to test block type ordering like this; love the legibility of it.

File: processor/parse/flavored/code-tabs.js [hunk 2]
```diff
@@ -1,50 +1,36 @@
-const RGXP = /^(```([^]*?)```)(?=\s*\n(?!```|\w)|```\n\n|$)/g;
+// eslint-disable-next-line unicorn/no-unsafe-regex
+const TAB_BLOCK_RGXP = /^(?:(?:^|\n)```(?:(?!\n```).)*\n```[^\S\n]*){2,}/gs;
+/*
+ * For each of our adjacent code blocks we'll split the matching block in to three parts:
+ *    - [lang] syntax extension
```
Comment (rafegoldberg) [inline processor/parse/flavored/code-tabs.js:5]:
```suggestion
 *    - [lang] syntax extension (optional)
```

File: __tests__/flavored-parsers.test.js [hunk 3]
```diff
@@ -99,4 +99,34 @@ describe('Parse RDMD Syntax', () => {
...
+  it('Allows trailing text on a closing fence.', () => {
+    const mdx = '```\nfoo\n```\n```\nbar\n``` Oops';
+    const ast = process(mdx);
+
+    expect(ast.children.map(c => c.type)).toStrictEqual(['code-tabs', 'paragraph']);
+  });
+
+  describe('Parsing individual code tabs', () => {
+    it('Handles triple backticks within a code block', () => {
+      const mdx = '```\nconsole.log("why would you do this?!```");\n```\n```\nbar\n```';
```
Comment (rafegoldberg) [inline __tests__/flavored-parsers.test.js:126]: lol, feel this. people are wont to do some crazy sh¡t with their markdown......

**Review by rafegoldberg — CHANGES_REQUESTED** [review]
<details>
<summary>Awesome work—I spent some time testing this PR against various <a href=https://rdmd.readme.io/docs/code-block-tests>scenarios</a> that have plagued us in the past, and it definitely fixes that pesky delimiter issue!</summary><hr>

    Section One
    ``` Plain
    console.log("zed");
    ```
    Section **Two**
    ```js Highlighted
    console.log("one");
    ```

Before | After
:--:|:--:
image | image

<hr></details>

My only hesitation is this 10.x failure—got a sense of what's going wrong? can we get it working before we merge this? Otherwise this looks great.

**Comment by kellyjosephprice** [comment]
> My only hesitation is this 10.x failure—got a sense of what's going wrong? can we get it working before we merge this? Otherwise this looks great.

If I remember correctly, named groups in `matchAll` isn't supported till 12.x. To keep the named groups, I can use `exec` in a `while` loop.

File: .eslintrc [hunk 4]
```diff
@@ -4,5 +4,8 @@
     "@readme/eslint-config/react"
   ],
   "parser": "babel-eslint",
-  "root": true
+  "root": true,
+  "rules": {
+    "no-cond-assign": ["error", "except-parens"]
```
Comment (kellyjosephprice) [inline .eslintrc:9]: @rafegoldberg does this seem fair?

Comment (rafegoldberg) [reply .eslintrc:9]: Shouldn't `"except-parens"` already be the default for this rule? Maybe we're overriding it in our core [ESLint config](https://github.com/readmeio/eslint-config)? (In which case we should update the rule in that repo too!) But yeah, I'm okay with this.

**Review by rafegoldberg — APPROVED** [review]
Nice.

**Comment by rafegoldberg** [comment]
## This PR was released!

🚀 Changes included in [<b>v6.22.1</b>](/readmeio/markdown/releases/tag/v6.22.1)
<br>

**Pull request merged**